src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,139 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Justin Thomas
+-/
+import data.polynomial
+import data.polynomial.ring_division
+import ring_theory.principal_ideal_domain
+import algebra.module.linear_map
+import field_theory.minpoly
+import linear_algebra
+import ring_theory.ideal.operations
+import ring_theory.polynomial_algebra
+
+/-!
+# Annihilating Ideal
+
+Given a commutative ring `R` and an `R`-algebra `A`
+Every element `a : A` defines
+an ideal `polynomial.annihilating_ideal a ⊆ R[X]`.
+Simply put, this is the set of polynomials `p` where
+the polynomial evaluation `p(a)` is 0.
+
+## Special case where the ground ring is a field
+
+In the special case that `R` is a field, we use the notation `R = 𝕜`.
+Here `𝕜[X]` is a PID, so there is a polynomial `g ∈ polynomial.annihilating_ideal a`
+which generates the ideal. We show that if this generator is
+chosen to be monic, then it is the minimal polynomial of `a`,
+as defined in `field_theory.minpoly`.
+
+## Special case: endomorphism algebra
+
+Given an `R`-module `M` (`[add_comm_group M] [module R M]`)
+there are some common specialiazations which may be more familiar.
+* Example 1: `A = M →ₗ[R] M`, the endomorphism algebra of an `R`-module M.
+* Example 2: `A = n × n` matrices with entries in `R`.
+-/
+
+open_locale polynomial
+
+namespace polynomial
+
+section general_case
+
+variables {R A : Type*} [comm_semiring R] [semiring A] [algebra R A]
+
+variables (R)
+
+/--
+The informal notation `p(a)` stand for `polynomial.aeval a p`.
+Again informally, the annihilating ideal of `a` is
+`{ p ∈ R[x] | p(a) = 0 }`. This is an ideal in `R[X]`.
+The formal definition uses the kernel of the aeval map. -/
+noncomputable def annihilating_ideal (a : A) : ideal R[X] :=
+ring_hom.ker (aeval a).to_ring_hom
+
+variables {R}
+
+/-- It is useful to refer to ideal membership sometimes
+ and the annihilation condition other times -/
+lemma mem_annihilating_ideal_iff_aeval_0 (a : A) (p : R[X]) :
+  p ∈ annihilating_ideal R a ↔ aeval a p = 0 :=
+iff.rfl
+
+end general_case
+
+section field_case
+
+variables {𝕜 A : Type*} [field 𝕜] [ring A] [algebra 𝕜 A]
+variables (𝕜)
+
+/-- Since `𝕜[x]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = annihilating_ideal a` -/
+noncomputable def annihilating_ideal_generator (a : A) : 𝕜[X] :=
+submodule.is_principal.generator (annihilating_ideal 𝕜 a)
+
+variables {𝕜}
+
+section minpoly_generates
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the first condition: it must annihilate
+the original element `a : A`. -/
+lemma annihilating_ideal_generator_aeval_0 (a : A) :
+  aeval a (annihilating_ideal_generator 𝕜 a) = 0 :=
+begin
+  rw annihilating_ideal_generator,
+  have gen_member := submodule.is_principal.generator_mem (annihilating_ideal 𝕜 a),
```

Comment:
if you're making `annihilating_ideal_generator` a def, then you should make a lemma for the fact it's a member too so that people that use `annihilating_ideal_generator` don't have to depend on the definition of `annihilating_ideal_generator` (what if you have to change it in future?)

Reply:
i see the point in this case. I am still to new at this to have an appreciation for the subtleties of how definitions can change over time.  Will add lemma for membership in next push.

Reply:
even if you don't expect the definition to ever change, it's still considered good form to do so